packages/amplify-e2e-tests/src/api-directives/common.ts

```diff
@@ -0,0 +1,366 @@
+    const subscriptionInput = testModule['input_' + subscriptionName];
+    const subscriptionResult = testModule['expected_result_' + subscriptionName];
+    const mutations = testModule['mutations_' + subscriptionName];
+    const mutationsInput = testModule['input_mutations_' + subscriptionName];
+
+    subscriptionTasks.push(async () => {
+      await testSubscription(appsyncClient, subscription, mutations, subscriptionResult, subscriptionInput, mutationsInput);
+    });
+  });
+
+  await runInSequential(subscriptionTasks);
+}
+
+export async function testSubscription(
+  appSyncClient: any,
+  subscription: string,
+  mutations: any[],
+  subscriptionResult: any,
+  subscriptionInput?: any,
+  mutationInputs?: any[],
+) {
+  const observer = appSyncClient.subscribe({
+    query: gql(subscription),
+    variables: subscriptionInput,
+  });
+
+  const received = [];
+  const sub = observer.subscribe((event: any) => {
+    received.push(event.data);
+  });
+
+  await new Promise(res => setTimeout(() => res(), 4000));
+
+  const mutationTasks = [];
+  for (let i = 0; i < mutations.length; i++) {
+    const mutation = mutations[i];
+    const mutationInput = mutationInputs ? mutationInputs[i] : undefined;
+    mutationTasks.push(async () => {
+      await appSyncClient.mutate({
+        mutation: gql(mutation),
+        fetchPolicy: 'no-cache',
+        variables: mutationInput,
+      });
+      await new Promise(res => setTimeout(() => res(), 4000)); //to ensure correct order in received data
+    });
+  }
+
+  await runInSequential(mutationTasks);
+
+  await new Promise(res => setTimeout(() => res(), 4000));
+
+  sub.unsubscribe();
+  if (!checkResult(received, subscriptionResult)) {
+    console.log('The following subscription test failed.');
+    console.log('Subscription: ', subscription);
+    if (subscriptionResult) {
+      console.log('Expected subscription result: ', subscriptionResult);
+    }
+    if (received) {
+      console.log('Actual subscription response: ', received);
+    }
+    throw new Error('Subscription test failed.');
+  }
+}
+
+/////
+function checkResult(received: any, expected: any): boolean {
+  if (!expected) {
+    //the test does not request result check, as long as the mutation/query goes through, it's good
+    return true;
+  }
+  const queue = [
+    {
+      received,
+      expected,
+      depth: 0,
+    },
+  ];
+  try {
+    return runCompare(queue);
+  } catch (e) {
+    console.log('checkResult error: ', e);
+    return false;
+  }
+}
+
+function checkError(received: any, expected: any): boolean {
+  if (!expected) {
+    //the test does not request result check, assume mutation/query should go through, but received error
+    return false;
+  }
+  const queue = [
+    {
+      received,
+      expected,
+      depth: 0,
+    },
+  ];
+  return runCompare(queue);
+}
+
+const MAX_DEPTH = 50;
+function runCompare(queue: { received: any; expected: any; depth: number }[]): boolean {
```

Comment:
Can we use [`.toMatchObject(object)`](https://jestjs.io/docs/en/expect#tomatchobjectobject) instead of rolling our own solution

Reply:
There are some technical difficulties. 
In the test, the actual received mutation, query and subscription results are not directly compared with the expected results specified in the test case file (e.g. using the *jest* *.toMatchObject(object)* method). 
- *id*, *createdAt* and *updatedAt* fields are generated (if not specified in the mutation), their values are not determined when the test starts to run. Those fields are generally given a value *“<check-defined>”* in the expected result specification, and the test only checks that they are defined in the actual received result. 
- If error is expected, the test only checks for the error type, but the actual error response contains a lot of fields.
- If actual response can contain the "__typename" field, even if the schema does not contain union of interface, and even if the mutation or query does not specify it, and the expected result does not include this extra field. 

To compare the expected and the actual result object, the test uses a BFS method to find mismatched field, if no mismatch is found after MAX_DEPTH (set at 50 in the test), the test passes.